### DISCUSSION TITLE
`brew uses` gives false-positives

### DISCUSSION THREAD

**Discussion by xxyxxyxyx1** [Homebrew/discussions#5786] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: 2e780004c92f16a96036d3d43569c2924c4e54f5
Last commit: 3 days ago
Branch: stable
Core tap JSON: 05 Dec 13:23 UTC
Core cask tap JSON: 05 Dec 13:23 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: ["--no-quarantine"]
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.OvOgTPJDCi/org.xquartz:0
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_NO_ENV_HINTS: set
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.6 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 16.0.0 build 1600
Git: 2.47.1 => /opt/homebrew/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 14.7.1-arm64
CLT: 16.1.0.0.1.1729049160
Xcode: 16.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1
```

### Description of issue

I wanted to check what formulae depend on `openssl@1.1` because it is deprecated, so I used:
```
brew uses --recursive --installed openssl@1.1
fftw grace scalapack
```

However, `brew info <formula>` does not show this dependency for those formulae. It seems `brew uses` gave false-positives:
```
brew info grace
==> grace: stable 5.1.25 (bottled)
WYSIWYG 2D plotting tool for X11
https://plasma-gate.weizmann.ac.il/Grace/
Installed
/opt/homebrew/Cellar/grace/5.1.25_5 (266 files, 9.5MB) *
  Poured from bottle using the formulae.brew.sh API on 2023-04-05 at 22:13:42
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/g/grace.rb
License: GPL-2.0-only
==> Dependencies
Required: fftw ✔, jpeg-turbo ✔, libice ✔, libpng ✔, libsm ✔, libx11 ✔, libxext ✔, libxmu ✔, libxp ✔, libxpm ✔, libxt ✔, openmotif ✔
==> Analytics
install: 111 (30 days), 376 (90 days), 1,646 (365 days)
install-on-request: 111 (30 days), 376 (90 days), 1,646 (365 days)
build-error: 0 (30 days)

brew info scalapack
==> scalapack: stable 2.2.0 (bottled)
High-performance linear algebra for distributed memory machines
https://netlib.org/scalapack/
Installed
/opt/homebrew/Cellar/scalapack/2.2.0_1 (25 files, 4.3MB) *
  Poured from bottle on 2022-08-06 at 12:35:55
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/s/scalapack.rb
License: BSD-3-Clause
==> Dependencies
Build: cmake ✔
Required: gcc ✔, open-mpi ✔, openblas ✔
==> Analytics
install: 205 (30 days), 593 (90 days), 2,218 (365 days)
install-on-request: 62 (30 days), 194 (90 days), 774 (365 days)
build-error: 1 (30 days)

brew info fftw
==> fftw: stable 3.3.10 (bottled)
C routines to compute the Discrete Fourier Transform
https://fftw.org
Installed
/opt/homebrew/Cellar/fftw/3.3.10_1 (74 files, 8.4MB) *
  Poured from bottle using the formulae.brew.sh API on 2024-12-05 at 14:27:07
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/f/fftw.rb
License: GPL-2.0-or-later AND BSD-2-Clause
==> Dependencies
Required: open-mpi ✔, gcc ✔
==> Analytics
install: 5,079 (30 days), 15,390 (90 days), 64,036 (365 days)
install-on-request: 643 (30 days), 1,981 (90 days), 8,649 (365 days)
build-error: 1 (30 days)
```

If I reinstall one of those formulae, such as `brew reinstall fftw`, then re-run the `brew uses` command, that freshly reinstalled formula is no longer a false-positive. 

In my case, none of the 5 formulae initially reported as depending on openssl@1.1 actually depend on it. 

Is this expected behavior of `brew uses`? Is it reflecting out-of-date dependency information? Could/should it use current dependency information?

I think it might be helpful to add something to the `brew uses --help` output regarding the accuracy of the information provided, if this is not simply a bug.

**Comment by gromgit** [comment, marked as answer]
What's the output of:
```
brew deps grace scalapack fftw
brew linkage grace scalapack fftw
brew ls --pinned
brew info libevent
```
? 

> However, `brew info <formula>` does not show this dependency for those formulae.

All three formulae you mentioned as false positives share an _indirect_ dependency on `libevent`, which in turn depended on `openssl@1.1` until 21 Jun 2023, when it was cut over to `openssl@3`. `brew info` only shows _direct_ dependencies.

> Is this expected behavior of brew uses? Is it reflecting out-of-date dependency information? Could/should it use current dependency information?

Not exactly out-of-date. `brew uses` usually looks at the JSON API (or local taps if `HOMEBREW_NO_INSTALL_FROM_API` is set) for the info it needs. However, you specified `brew uses --installed` which instead looks at the (recursive) dependencies recorded in the manifest of each formula _at the time you installed it_. Your `brew info` output for both `grace` and `scalapack` clearly indicate installation prior to `libevent`'s cutover to `openssl@3`.

**Reply by xxyxxyxyx1** [reply]
Here is output:
```      
brew deps grace scalapack fftw
ca-certificates
gcc
gmp
hwloc
isl
libevent
libmpc
lz4
mpfr
open-mpi
openssl@3
pmix
xz
zstd

no output for brew ls --pinned

brew info libevent
==> libevent: stable 2.1.12 (bottled)
Asynchronous event library
https://libevent.org/
Installed
/opt/homebrew/Cellar/libevent/2.1.12_1 (57 files, 2.2MB) *
  Poured from bottle using the formulae.brew.sh API on 2023-06-29 at 01:51:09
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/lib/libevent.rb
License: BSD-3-Clause
==> Dependencies
Build: autoconf ✔, automake ✔, libtool ✔, pkgconf ✔
Required: openssl@3 ✔
==> Analytics
install: 54,332 (30 days), 174,529 (90 days), 812,760 (365 days)
install-on-request: 910 (30 days), 3,124 (90 days), 19,842 (365 days)
build-error: 1 (30 days)
```

note: I mentioned in my question that I resolved my situation with `brew reinstall`; thus, the output I just pasted here might be no longer applicable. If there is some quick method to revert to my prior state for testing, I would be happy to do that.

**Reply by gromgit** [reply]
All the output you've shown so far is in fact consistent with what I said:

> you specified `brew uses --installed` which instead looks at the (recursive) dependencies recorded in the manifest of each formula _at the time you installed it_

Note I said "recursive". Every formula's manifest records all direct _and_ indirect dependencies at the time of installation. Since your `libevent` _and_`fftw` were (re)installed after the former's move to `openssl@3`, both their manifests no longer record `openssl@1.1` as a dependency, direct (`libevent`) or indirect (`fftw`, as you noted, no longer appears in the output of `brew uses --installed openssl@1.1`). Reinstalling `scalapack` and `grace` would have the same effect.
